capnp-rpc/src/local.rs

```diff
@@ -368,15 +368,38 @@ where
         let inner = self.inner.clone();
         Promise::from_future(async move {
             let f = {
+                let this_hook = Self { inner };
+
+                // We need to provide a way to get the corresponding ClientHook of a Server.
+                // Passing the ClientHook inside `Params` would require a lot of changes, breaking compatiblity
+                // with existing code and ruining the developer experience, because `Params` would end up containing more generics.
```

Comment:
What if `this_client()` returned an untyped capability -- perhaps just a `Box<dyn ClientHook>`? That would not require any more generics, would it?
(It looks to me like that's how capnproto-c++ implements `thisCap()`.)

Reply:
Having a `params.this_client()` returning `Box<dyn ClientHook>` would be possible but it has the following drawbacks:
- the user needs to manually cast the client. 
- `params` is passed by value, and can be moved around and stored eventually for a `'static` lifetime. That means `params` would need to own a copy of the current client, requiring a `Box` allocation for each method call. Unless we store a `Weak` reference to the `ClientHook`, but then `params.this_client()` needs to return an `Option<T>`, because the `params` we own may reference a dead `ClientHook`.

Returning an `Option<Box<dyn ClientHook>>` isn't great compared to a fully typed `calculator::op_add::Client`

Reply:
An alternative API which doesn't require unsafe nor a static variable is
`params.client_for(self)`. (Where self is actually `&mut self`, because we are inside a method call).

By requiring `&mut self` as a parameter, we restrict the ability to get a client while we are  inside the method call, and we also have the type information required to return a typed Client. 

It's fully safe and typed Rust.